sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala

```diff
@@ -236,6 +236,45 @@ case class ShowFunctionsCommand(
   }
 }
 
+
+/**
+ * A command for users to refresh the persistent function.
+ * The syntax of using this command in SQL is:
+ * {{{
+ *    REFRESH FUNCTION functionName
+ * }}}
+ */
+case class RefreshFunctionCommand(
+    databaseName: Option[String],
+    functionName: String)
+  extends RunnableCommand {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val catalog = sparkSession.sessionState.catalog
+    if (FunctionRegistry.builtin.functionExists(FunctionIdentifier(functionName))) {
+      throw new AnalysisException(s"Cannot refresh builtin function $functionName")
+    }
+    if (catalog.isTemporaryFunction(FunctionIdentifier(functionName, databaseName))) {
+      throw new AnalysisException(s"Cannot refresh temporary function $functionName")
+    }
+
+    val identifier = FunctionIdentifier(
+      functionName, Some(databaseName.getOrElse(catalog.getCurrentDatabase)))
+    // we only refresh the permanent function.
+    if (catalog.isPersistentFunction(identifier)) {
+      // register overwrite function.
+      val func = catalog.getFunctionMetadata(identifier)
+      catalog.registerFunction(func, true)
+    } else {
+      // function is not exists, clear cached function.
+      catalog.unregisterFunction(identifier, true)
+      throw new NoSuchFunctionException(identifier.database.get, functionName)
```

Comment:
If it's a valid use case, why do we throw an exception here?

Reply:
Just keep the same behavior with `refresh table`, the later also throw `NoSuchTableException`.

Reply:
are you sure `REFRESH TABLE` can do some work which has side-effects and still fail?

Reply:
`REFRESH TABLE` doesn't do the side-effects, it always check the table if exist first.

I think it's necessary to have both of invalid cache and throw exception.
* It's confused that we can still use or desc a not exist function if we just throw exception. 
* It's also confused that we can refresh any function name without an exception if we just clear cache.

BTW current `REFRESH TABLE` exists a minor memory leak in this case
```
-- client a execute
create table t1(c1 int);
cache table t1;

-- client b execute
drop table t1;
create table t1(c1 int, c2 int);
uncache table t1.

-- client a.t1 produce a memory leak
-- the reason is spark think it's a plan cache but user may think it's a table cache
```

Reply:
It depends on how you define "function exists". If users can still use this function in SQL queries, why do we throw NoSuchFunctionException in REFRESH FUNCTION?

Reply:
how about this 
```
if (catalog.isPersistentFunction(identifier)) {
  // register overwrite function.
  val func = catalog.getFunctionMetadata(identifier)
  catalog.registerFunction(func, true)
} else if (catalog.isRegisteredFunction(identifier)) {
  // clear cached function.
  catalog.unregisterFunction(identifier, true)
} else {
  throw new NoSuchFunctionException(identifier.database.get, functionName)
}
```

Reply:
LGTM